File: platform/polyglot/healthz_diagnostics/py/healthz_diagnostics/executor.py
import asyncio
from typing import Any, Dict, Protocol, Optional, Callable
from .logger import create as create_logger
from .collector import DiagnosticsCollector
from .timestamp import TimestampFormatter
from .types import HealthCheckResult, ProviderConfig

logger = create_logger("healthz_diagnostics", __name__)

class HttpClient(Protocol):
    async def request(self, method: str, url: str) -> Any: ...
    async def close(self) -> None: ...

HttpClientFactory = Callable[[Dict[str, Any]], HttpClient]

class HealthCheckExecutor:
    """
    Orchestrates health checks with diagnostics, timing, and error handling.
    """

    def __init__(self, http_client_factory: HttpClientFactory, logger_instance: Optional[Any] = None):
        self._client_factory = http_client_factory
        self._logger = logger_instance or logger
        self._timestamp = TimestampFormatter()

    async def execute(self, provider_name: str, provider_config: Dict[str, Any]) -> HealthCheckResult:
        """
        Execute health check for a single provider.
        """
        self._logger.info(f"Executing health check for {provider_name}")
        
        collector = DiagnosticsCollector()
        status_code: Optional[int] = None
        error_msg: Optional[str] = None
        healthy = False
        response_data: Optional[Any] = None
        
        base_url = provider_config.get("base_url")
        health_endpoint = provider_config.get("health_endpoint", "")
        model = provider_config.get("model")
        method = (provider_config.get("method") or "GET").upper()
        health_body = provider_config.get("health_body")
        
        if not base_url:
            error_msg = f"{provider_name} provider not configured (missing base_url)"
            collector.push_error(error_msg)
            return self._build_result(
                provider_name, False, None, error_msg, None, model, collector, None
            )

        # URL construction per REQ0006
        # Always remove trailing slash from base_url
        clean_base = base_url.rstrip("/")

        if not health_endpoint:
            full_url = clean_base
            normalized_endpoint = ""
        else:
            # Ensure health_endpoint starts with /
            normalized_endpoint = health_endpoint if health_endpoint.startswith("/") else f"/{health_endpoint}"
            full_url = f"{clean_base}{normalized_endpoint}"

        client = self._client_factory(provider_config)

        try:
            collector.push_start(full_url, method)

            # Execute request - pass normalized path, not full URL
            # (client already has base_url configured)
            if health_body and method == "POST":
                response = await client.request(method, normalized_endpoint, json=health_body)
            else:
                response = await client.request(method, normalized_endpoint)
            
            # Handle response objects that might be dict or object (depending on client)
            # Assuming client returns object with status_code or dict with status
            if isinstance(response, dict):
                status_code = response.get("status", response.get("status_code"))
                response_data = response.get("data")
            else:
                status_code = getattr(response, "status_code", getattr(response, "status", None))
                response_data = getattr(response, "data", None)

            if status_code and 200 <= status_code < 300:
                healthy = True
                collector.push_end(status_code)
            else:
                healthy = False
                error_msg = f"Unhealthy status code: {status_code}"
                collector.push_error(error_msg)

        except Exception as e:
            healthy = False
            error_msg = str(e)
            collector.push_error(error_msg)
            self._logger.error(f"Health check failed for {provider_name}: {e}")
            
        finally:
            try:
                await client.close()
            except Exception as close_err:
                self._logger.warn(f"Failed to close client for {provider_name}: {close_err}")

        return self._build_result(
            provider_name,
            healthy,
            status_code,
            error_msg,
            full_url,
            model,
            collector,
            response_data
        )

    def _build_result(
        self,
        provider: str,
        healthy: bool,
        status_code: Optional[int],
        error: Optional[str],
        endpoint: Optional[str],
        model: Optional[str],
        collector: DiagnosticsCollector,
        data: Optional[Any]
    ) -> HealthCheckResult:
        return {
            "provider": provider,
            "healthy": healthy,
            "status_code": status_code,
            "latency_ms": collector.get_duration() * 1000,
            "error": error,
            "endpoint": endpoint,
            "model": model,
            "timestamp": self._timestamp.format(),
            "diagnostics": collector.get_events(),
            "data": data
        }

File: platform/polyglot/healthz_diagnostics/mjs/src/executor.mjs
import { create as createLogger } from "./logger.mjs";
import { DiagnosticsCollector } from "./collector.mjs";
import { TimestampFormatter } from "./timestamp.mjs";

const defaultLogger = createLogger("healthz-diagnostics", import.meta.url);

/**
 * Orchestrates health checks with diagnostics, timing, and error handling.
 */
export class HealthCheckExecutor {
  #clientFactory;
  #logger;
  #timestamp;

  /**
   * @param {Function} httpClientFactory - Factory that returns an HTTP client
   * @param {object} [logger] - Optional logger override
   */
  constructor(httpClientFactory, logger = null) {
    this.#clientFactory = httpClientFactory;
    this.#logger = logger || defaultLogger;
    this.#timestamp = new TimestampFormatter();
  }

  /**
   * Execute health check for a single provider.
   * @param {string} providerName
   * @param {object} providerConfig
   * @returns {Promise<import('../types.d.ts').HealthCheckResult>}
   */
  async execute(providerName, providerConfig) {
    this.#logger.info(`Executing health check for ${providerName}`);

    const collector = new DiagnosticsCollector();
    let statusCode = null;
    let errorMsg = null;
    let healthy = false;
    let responseData = null;

    const baseUrl = providerConfig.base_url;
    const healthEndpoint = providerConfig.health_endpoint || "";
    const model = providerConfig.model;
    const method = (providerConfig.method || "GET").toUpperCase();
    const healthBody = providerConfig.health_body || null;

    if (!baseUrl) {
      errorMsg = `${providerName} provider not configured (missing base_url)`;
      collector.pushError(errorMsg);
      return this.#buildResult(
        providerName,
        false,
        null,
        errorMsg,
        null,
        model,
        collector,
        null
      );
    }

    // URL construction per REQ0006
    // Always remove trailing slash from baseUrl
    const cleanBase = baseUrl.endsWith("/") ? baseUrl.slice(0, -1) : baseUrl;
    let fullUrl;
    let normalizedEndpoint;

    if (!healthEndpoint) {
      fullUrl = cleanBase;
      normalizedEndpoint = "";
    } else {
      // Ensure healthEndpoint starts with /
      normalizedEndpoint = healthEndpoint.startsWith("/")
        ? healthEndpoint
        : `/${healthEndpoint}`;
      fullUrl = `${cleanBase}${normalizedEndpoint}`;
    }

    const client = this.#clientFactory(providerConfig);

    try {
      collector.pushStart(fullUrl, method);

      // Execute request - pass normalized path, not full URL
      // (client already has base_url configured)
      const requestOpts = { method, path: normalizedEndpoint };
      if (healthBody && method === "POST") {
        requestOpts.json = healthBody;
      }
      const response = await client.request(requestOpts);

      statusCode =
        response.status || response.statusCode || response.status_code;
      responseData = response.data;

      if (statusCode && statusCode >= 200 && statusCode < 300) {
        healthy = true;
        collector.pushEnd(statusCode);
      } else {
        healthy = false;
        errorMsg = `Unhealthy status code: ${statusCode}`;
        collector.pushError(errorMsg);
      }
    } catch (e) {
      healthy = false;
      errorMsg = e instanceof Error ? e.message : String(e);
      collector.pushError(errorMsg);
      this.#logger.error(`Health check failed for ${providerName}: ${e}`);
    } finally {
      try {
        if (client.close) {
          await client.close();
        }
      } catch (closeErr) {
        this.#logger.warn(
          `Failed to close client for ${providerName}: ${closeErr}`
        );
      }
    }

    return this.#buildResult(
      providerName,
      healthy,
      statusCode,
      errorMsg,
      fullUrl,
      model,
      collector,
      responseData
    );
  }

  #buildResult(
    provider,
    healthy,
    statusCode,
    error,
    endpoint,
    model,
    collector,
    data
  ) {
    return {
      provider,
      healthy,
      status_code: statusCode,
      latency_ms: collector.getDuration() * 1000,
      error,
      endpoint,
      model,
      timestamp: this.#timestamp.format(),
      diagnostics: collector.getEvents(),
      data,
    };
  }
}
